packages/plugin-node-modules/sources/NodeModulesLinker.ts

```diff
@@ -664,9 +793,49 @@ async function persistNodeModules(preinstallState: NodeModulesLocatorMap | null,
     await Promise.all(addQueue);
 
     await xfs.mkdirpPromise(rootNmDirPath);
-    await writeInstallState(project, installState);
+
+    const binSymlinks = await createBinSymlinkMap(installState, locationTree, {loadManifest});
+    await persistBinSymlinks(prevBinSymlinks, binSymlinks);
+
+    await writeInstallState(project, installState, binSymlinks);
   } finally {
     reportedProgress.stop();
   }
 };
 
+async function persistBinSymlinks(previousBinSymlinks: BinSymlinkMap, binSymlinks: BinSymlinkMap) {
+  // Delete outdated .bin folders
+  for (const location of previousBinSymlinks.keys()) {
+    if (!binSymlinks.has(location)) {
+      const binDir = ppath.join(location, NODE_MODULES, DOT_BIN);
+      xfs.removeSync(binDir);
```

Comment:
Just to be sure, do you really want to use the sync variants, rather than promises?

Reply:
They were a little bit faster initially when I had dumber algorithm, but now the algorithm has fewer IO ops and in any case itts certainly not a bottleneck, lets use promises for uniformity.